streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -463,9 +464,8 @@ private void replaceStreamThread(final Throwable throwable) {
             closeToError();
         }
         final StreamThread deadThread = (StreamThread) Thread.currentThread();
-        threads.remove(deadThread);
```

Comment:
We cannot wait here until the dead thread is shutdown because the shutdown happens after `replaceStreamThread()` throws the exception. So we would wait forever.

Reply:
`deadThread.shutdown();` I was referring to this below. But if we don't need to keep the same for any reason I am fine either way

Reply:
Sorry for not being clear enough. I was also referring to `deadThread.shutdown()`. Method `deadThread.shutdown()` only requests a shutdown. The actual shutdown is performed in `completeShutdown()` which is called after `replaceStreamThread()` throws one of the exceptions below. Since `completeShutdown()` is called by the same thread that calls `deadThread.shutdown()` in this method, i.e., the dead thread, we would wait forever if we waited after `deadThread.shutdown()` until the dead thread is shut down. Or did I misunderstand your statement "by waiting a bit in the replace thread".

Reply:
Ah that makes sense. Another approach is using a temporary name. And then waiting the new thread until the old thread dies and takes the name. This is a bit complicated and I think it should only be done if it is necessary for the new thread to have the same name. And probably not in this PR but it could be an improvement done later

Reply:
Yeah I think swapping the names would make the code unnecessarily complicated, and it would definitely make reading the logs more difficult. 
Just to note: in the current rebalance protocol, the thread name should not impact the task assignment since within a client tasks are always just assigned to their previous owner (we maximize stickiness & balance)

Reply:
since the name doesn't matter I wonder why we spent so much effort making sure it had the same name?